passes/techmap/clkbufmap.cc

```diff
@@ -0,0 +1,299 @@
+
+		if (buf_celltype.empty())
+			log_error("The -buf option is required.");
+
+		// Cell type, port name, bit index.
+		pool<pair<IdString, pair<IdString, int>>> sink_ports;
+		pool<pair<IdString, pair<IdString, int>>> buf_ports;
+
+		// Process submodules before module using them.
+		std::vector<Module *> modules_sorted;
+		pool<Module *> modules_processed;
+		for (auto module : design->selected_modules())
+			module_queue(design, module, modules_sorted, modules_processed);
+
+		for (auto module : modules_sorted)
+		{
+			if (module->get_blackbox_attribute()) {
+				auto it = module->attributes.find("\\clkbuf_driver");
+				if (it != module->attributes.end()) {
+					auto value = it->second.decode_string();
+					for (auto name : split(value, ",")) {
+						auto wire = module->wire(RTLIL::escape_id(name));
+						if (!wire)
+							log_error("Module %s does not have port %s.\n", log_id(module), log_id(name));
+						for (int i = 0; i < GetSize(wire); i++)
+							buf_ports.insert(make_pair(module->name, make_pair(RTLIL::escape_id(name), i)));
+					}
+				}
+				it = module->attributes.find("\\clkbuf_sink");
+				if (it != module->attributes.end()) {
+					auto value = it->second.decode_string();
+					for (auto name : split(value, ",")) {
+						auto wire = module->wire(RTLIL::escape_id(name));
+						if (!wire)
+							log_error("Module %s does not have port %s.\n", log_id(module), log_id(name));
+						for (int i = 0; i < GetSize(wire); i++)
+							sink_ports.insert(make_pair(module->name, make_pair(RTLIL::escape_id(name), i)));
+					}
+				}
+				continue;
+			}
+			pool<SigBit> sink_wire_bits;
+			pool<SigBit> buf_wire_bits;
+			pool<SigBit> driven_wire_bits;
+			SigMap sigmap(module);
+			// bit -> (buffer, buffer's input)
+			dict<SigBit, pair<Cell *, Wire *>> buffered_bits;
+
+			// First, collect nets that could use a clock buffer.
+			for (auto cell : module->cells())
+			for (auto port : cell->connections())
+			for (int i = 0; i < port.second.size(); i++)
+				if (sink_ports.count(make_pair(cell->type, make_pair(port.first, i))))
+					sink_wire_bits.insert(sigmap(port.second[i]));
+
+			// Second, collect ones that already have a clock buffer.
+			for (auto cell : module->cells())
+			for (auto port : cell->connections())
+			for (int i = 0; i < port.second.size(); i++)
+				if (buf_ports.count(make_pair(cell->type, make_pair(port.first, i))))
+					buf_wire_bits.insert(sigmap(port.second[i]));
+
+			// Collect all driven bits.
+			for (auto cell : module->cells())
+			for (auto port : cell->connections())
+				if (cell->output(port.first))
+					for (int i = 0; i < port.second.size(); i++)
+						driven_wire_bits.insert(port.second[i]);
+
+			// Insert buffers.
+			std::vector<pair<Wire *, Wire *>> input_queue;
+			for (auto wire : module->selected_wires())
+			{
+				// Should not happen.
+				if (wire->port_input && wire->port_output)
+					continue;
+				if (wire->get_bool_attribute("\\clkbuf_inhibit"))
```

Comment:
Perhaps we can make this a user defined option given to the `clkbufmap` pass? That way it can be generic to match whatever the style guide or vendor rules exist

Reply:
That'd be a useful feature, but wouldn't it be better to use the generic `attrmap` pass instead of coding it in here?

Reply:
Absolutely possible through `attrmap`, but that's another walk over the design when you're doing one here anyway

Reply:
Another reason for a user specified attribute: https://github.com/YosysHQ/nextpnr/pull/317 as opposed to a hardcoded one (and to reduce the number of hardcoded attributes Yosys has to support)